snuba/utils/streams/dummy.py

```diff
@@ -102,90 +149,90 @@ def subscribe(
     ) -> None:
-        if self.__closed:
-            raise RuntimeError("consumer is closed")
+        with self.__lock:
+            if self.__closed:
+                raise RuntimeError("consumer is closed")
 
-        # TODO: Handle offset reset more realistically.
-        self.__offsets = {
-            partition: self.__broker.offsets[self.__group].get(partition, 0)
-            for partition in self.__broker.subscribe(self, topics)
-        }
+            self.__offsets = {**self.__broker.subscribe(self, topics)}
 
-        self.__staged_offsets.clear()
-        self.__last_eof_at.clear()
+            self.__staged_offsets.clear()
+            self.__last_eof_at.clear()
 
-        if on_assign is not None:
-            on_assign(self.__offsets)
+            if on_assign is not None:
+                on_assign(self.__offsets)
 
     def unsubscribe(self) -> None:
-        if self.__closed:
-            raise RuntimeError("consumer is closed")
+        with self.__lock:
+            if self.__closed:
+                raise RuntimeError("consumer is closed")
 
-        self.__broker.unsubscribe(self)
+            self.__broker.unsubscribe(self)
 
-        self.__offsets = {}
-        self.__staged_offsets.clear()
-        self.__last_eof_at.clear()
+            self.__offsets = {}
+            self.__staged_offsets.clear()
+            self.__last_eof_at.clear()
 
     def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
-        if self.__closed:
-            raise RuntimeError("consumer is closed")
+        with self.__lock:
+            if self.__closed:
+                raise RuntimeError("consumer is closed")
 
-        # TODO: Throw ``EndOfPartition`` errors.
-        for partition, offset in sorted(self.__offsets.items()):
-            if partition in self.__paused:
-                continue  # skip paused partitions
+            for partition, offset in sorted(self.__offsets.items()):
+                if partition in self.__paused:
+                    continue  # skip paused partitions
 
-            messages = self.__broker.topics[partition.topic][partition.index]
-            try:
-                payload = messages[offset]
-            except IndexError:
-                if offset == len(messages):
+                try:
+                    message = self.__broker.consume(partition, offset)
+                except Exception as e:
+                    raise ConsumerError("error consuming mesage") from e
+
+                if message is None:
                     if self.__enable_end_of_partition and (
                         partition not in self.__last_eof_at
                         or offset > self.__last_eof_at[partition]
                     ):
                         self.__last_eof_at[partition] = offset
                         raise EndOfPartition(partition, offset)
                 else:
-                    raise ConsumerError("invalid offset")
```

Comment:
This is now an Exception and not a ConsumerError. Is this intended ?

Reply:
This is still raised as a `ConsumerError` from the consumer, as I felt it did not make much sense for the broker to raise `ConsumerError`: https://github.com/getsentry/snuba/blob/afd82790a8209d1f8c8256260ad79f267d6cfb0d/snuba/utils/streams/dummy.py#L187

This is tested here: https://github.com/getsentry/snuba/blob/afd82790a8209d1f8c8256260ad79f267d6cfb0d/tests/utils/streams/mixins.py#L229-L243